standalone/pruntime/enclave/src/libc_hacks.rs

```diff
@@ -0,0 +1,424 @@
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn writev(fd: c_int, iov: *const iovec, iovcnt: c_int) -> ssize_t {
+    assert_eq_size!(libc::iovec, sgx_libc::iovec);
+
+    if !matches!(fd, 1 | 2) {
+        not_allowed!()
+    } else {
+        unsafe { ocall::writev(fd, iov as _, iovcnt) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn getcwd(buf: *mut c_char, size: size_t) -> *mut c_char {
+    // Enclave have no working directory, let's return ""
+    if size > 0 {
+        unsafe { *buf = 0 };
+    }
+    return buf;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn syscall(num: libc::c_long, mut args: ...) -> libc::c_long {
+    if num == libc::SYS_getrandom {
+        let buf = args.arg::<*mut c_void>();
+        let buflen = args.arg::<size_t>();
+        let flags = args.arg::<c_uint>();
+        return getrandom(buf, buflen, flags) as _;
+    }
+    eprintln!("unsupported syscall({})", num);
+    loop {}
+}
+
+#[no_mangle]
+pub extern "C" fn memrchr(s: *mut c_void, c: c_int, n: size_t) -> *mut c_void {
+    unsafe { sgx_libc::memrchr(s as _, c as _, n) as _ }
+}
+
+// sgx_mutex will ignore attributes
+#[no_mangle]
+pub extern "C" fn pthread_mutexattr_init(_attr: *mut libc::pthread_mutexattr_t) -> c_int {
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn pthread_mutexattr_settype(
+    _attr: *mut libc::pthread_mutexattr_t,
+    typ: c_int,
+) -> c_int {
+    // sgx pthread_mutex only support normal mutex, but the stdout() requires PTHREAD_MUTEX_RECURSIVE type.
+    // assert_eq!(typ, libc::PTHREAD_MUTEX_NORMAL);
+
+    static RE: AtomicU16 = AtomicU16::new(0);
+
+    if typ != libc::PTHREAD_MUTEX_NORMAL {
+        // workaround for stdout(), stderr() and stdin()
+        let count = RE.fetch_add(1, Ordering::Relaxed);
+        if count > 3 {
+            panic!("only PTHREAD_MUTEX_NORMAL supported");
+        }
+    }
+
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn pthread_mutexattr_destroy(_attr: *mut libc::pthread_mutexattr_t) -> c_int {
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn sched_yield() -> c_int {
+    unsafe { ocall::sched_yield() }
+}
+
+extern "C" {
+    fn strerror_r(errnum: c_int, buf: *mut c_char, buflen: size_t) -> c_int;
+}
+
+#[no_mangle]
+pub extern "C" fn __xpg_strerror_r(errnum: c_int, buf: *mut c_char, buflen: size_t) -> c_int {
+    unsafe { strerror_r(errnum, buf, buflen) }
+}
+
+#[no_mangle]
+pub extern "C" fn clock_gettime(clk_id: libc::clockid_t, tp: *mut libc::timespec) -> c_int {
+    assert_eq_size!(libc::timespec, sgx_libc::timespec);
+
+    unsafe { ocall::clock_gettime(clk_id, tp as _) }
+}
+
+#[no_mangle]
+pub extern "C" fn close(_fd: c_int) -> c_int {
+    not_allowed!()
+}
+
+#[no_mangle]
+pub extern "C" fn getrandom(buf: *mut c_void, buflen: size_t, _flags: c_uint) -> ssize_t {
+    if buflen == 0 {
+        return 0;
+    }
+
+    let rv = unsafe { sgx_read_rand(buf as _, buflen) };
+
+    match rv {
+        sgx_status_t::SGX_SUCCESS => buflen as _,
+        _ => {
+            sgx_libc::set_errno(libc::EINTR);
```

Comment:
Just curious, why do you choose EINTR as the error message here?

Reply:
sgx_read_rand might fail due to `hardware not ready`. Caller beleave it-self is under linux, so EINTR means the syscall is interrupted be a signal, then caller might catch this error and try again later.

Reply:
Ack